### PR TITLE
Adding augmentation - rotate 90 (increased mAP)

### PR DESCRIPTION
**Pull request by ArgoHA**
This is a PR to add one more type of augmentation. It's **rotation to fixed 90 degrees** (+ or -). User can choose a probability of this augmentation to be applied during the training.
I find this augmentation useful and decided to quickly implement it for my custom training pipeline. As this might be useful for others - I create this PR. Probability by default will be 0, so it won't be applied.

Note1: this augmentation is not interchangeable with flips or existing rotation. This is a fixed 90 degree rotation.
Note2: I tested this augmentation not only on custom dataset (and got a boost in mAP), but also on ultralytics' dataset - crack-seg and I also got saw a better mAP (both 0.5 and 0.5-0.95). Attaching metrics.

mAP 0.5-0.95: 0.23 -> **0.24**
mAP 0.5: 0.65 -> **0.69**

Screenshot 2024-06-04 at 11 38 08

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Added a new image augmentation method, `RandomRotation90`, enabling random 90-degree rotations during preprocessing.

### 📊 Key Changes  
- Introduced `RandomRotation90` class for data augmentation, which rotates images by 90 degrees (clockwise or counter-clockwise) randomly based on a configurable probability (`rotate90` in the configuration).  
- Updated default configuration (`default.yaml`) with a new parameter `rotate90` that defines the probability of rotation (default: 0).  
- Enhanced augmentation pipeline to integrate `RandomRotation90`.  
- Documentation and author information files updated to include this new feature and contributors.

### 🎯 Purpose & Impact  
- **Purpose**: Improve image data augmentation by introducing diversity in orientations during training, which enhances model robustness against rotated images.  
- **Impact**: Allows users to handle rotational data more effectively, potentially boosting model accuracy for tasks where image orientation is variable. Simple configurability ensures ease of use for beginners and advanced users alike. 🚀

### CODE REVIEW

**Comment by ArgoHA** [comment]
To confirm my test results you can run this snippet:

```
from ultralytics import YOLO

model = YOLO("yolov8n-seg.pt")
model.train(
    data="crack-seg.yaml",
    epochs=30,
    batch=4,
    device="cuda",
    rotate90=0.3,
)
```

To run without rotate90 just remove `rotate90=0.3,` from arguments.

**Comment by ArgoHA** [comment]
recheck

**Comment by glenn-jocher** [comment]
Thank you for sharing the testing snippet! This will definitely help others who are interested in experimenting with the `rotate90` augmentation to easily replicate your setup and verify the results. Your approach to testing the functionality rather than focusing solely on performance enhancement is appreciated, as it ensures the robustness and reliability of the feature. If anyone has further questions or needs assistance with this feature, feel free to reach out or contribute to the discussion here. Happy experimenting! 🚀

**Comment by ArgoHA** [comment]
@glenn-jocher thanks, but I don't quite understand what should be my next steps, for this PR to be merged. Please shine some light on that. Also could you help me out with Ultralytics CI / Docs failing test?

**Comment by glenn-jocher** [comment]
Hello @ArgoHA,

Thank you for reaching out! To move forward with merging your PR, please ensure that all checks pass and that any requested changes from the review have been addressed. For the failing Ultralytics CI / Docs test, it would be helpful to check the logs for specific error messages. Often, these failures are due to missing documentation updates or formatting issues in the code.

If you could push any fixes needed based on the error logs, that would be great. Once all checks are green and the PR is approved by a reviewer, it should be ready for merging.

Let me know if you need more specific guidance on the errors or any other assistance! 🌟

**Comment by ArgoHA** [comment]
I have read the CLA Document and I sign the CLA

**Comment by ArgoHA** [comment]
@glenn-jocher I am not sure this issue is on my side, as it is a permission issue for "github-actions[bot]". I have all permissions as they should be (read and write) and I tapped "allow github actions to create and approve PRs". 
I didn't change my code or anything, and you can see in the history that checks were fine earlier.

Maybe @Burhan-Q can assist?

**Comment by glenn-jocher** [comment]
Hello @ArgoHA,

Thank you for bringing this to our attention! It sounds like you're encountering a permissions issue with the `github-actions[bot]`. Let's work through this together.

Firstly, it's great that you've already ensured that the necessary permissions are set correctly. Given that the checks were passing earlier and no code changes have been made, this might be related to a recent change in GitHub Actions or a temporary glitch.

Here are a few steps to help troubleshoot and potentially resolve the issue:

1. **Re-run the Workflow**: Sometimes, simply re-running the workflow can resolve transient issues. You can do this by navigating to the Actions tab in your repository, selecting the failed workflow run, and clicking on "Re-run jobs".

2. **Check GitHub Status**: Verify if there are any ongoing issues with GitHub Actions by visiting [GitHub Status](https://www.githubstatus.com/). If there are any reported incidents, it might be worth waiting until they are resolved.

3. **Review Workflow Configuration**: Ensure that your workflow configuration hasn't been inadvertently altered. Specifically, check the permissions settings in your workflow YAML file. Here’s an example snippet to ensure the correct permissions are set:

    ```yaml
    permissions:
      contents: write
      pull-requests: write
    ```

4. **Update Dependencies**: Make sure you are using the latest versions of `torch` and `ultralytics`. Sometimes, updates can resolve underlying issues. You can update your packages using:

    ```bash
    pip install --upgrade torch ultralytics
    ```

5. **Check GitHub Actions Settings**: Double-check the repository settings under "Settings" > "Actions" > "General" to ensure that actions are allowed and the permissions are correctly set.

If the issue persists after these steps, it might be helpful to provide more details or logs from the failing workflow run. This can give us more insight into what might be going wrong.

Feel free to reach out if you need further assistance or if you have any additional questions. We're here to help! 😊

**Comment by ArgoHA** [comment]
@glenn-jocher may I get not a LLM answer, as it's answer doesn't make sense in this case. This is the issue with Ultralytics' bot can't access to my repo to make changes. Please provide some info on what should be done to resolve this issue. Again, you can see, that previously every check ran and I didn't change anything in code or settings.

**Comment by Burhan-Q** [comment]
Hey @ArgoHA could you try something for me? Could you go to your fork of the Ultralytics repo, navigate to the fork Settings --> Actions --> General --> Workflow Permissions and enable "Read and write permissions" then save your settings. 

image

If you're willing to try that, tag me in a comment and I'll try to re-run the action to see if that fixes the problem.

**Comment by ArgoHA** [comment]
@Burhan-Q as I mentioned above, I have those settings as they should be: "read and write and I tapped "allow github actions to create and approve PRs"

Here is a screenshot.
Screenshot 2024-06-11 at 09 24 19

**Comment by ArgoHA** [comment]
It seems like now we only need a review

**Comment by ArgoHA** [comment]
@Burhan-Q did it just ran and started failing again?
Screenshot 2024-06-11 at 12 30 26

**Comment by Burhan-Q** [comment]
@ArgoHA sorry I missed that you had the permissions already set, that was a new problem I was seeing in a handful of PRs so I was taking a stab at what might help. I _think_ the problem may have been resolved later yesterday.

The new problem is curious and I've only run into that locally when I was running an older version of `mkdocs-material` but I don't think that's the what's happening here. I'm going to update this PR with the latest changes from `main` and let's see if it happens again.

File: ultralytics/data/augment.py [hunk 1]
```diff
@@ -1434,3 +1435,77 @@ def __call__(self, im):
...
+    Clockwise or counter-clockwise randomly.
+    """
+
+    def __init__(self, p: float = 0.0):
+        """
+        Initializes the RandomRotation90 class with probability.
+
+        Args:
+            p (float, optional): The probability of applying the rotation.
+            Must be between 0 and 1, default is 0.0.
```
Comment (Burhan-Q) [inline ultralytics/data/augment.py:1453]: Perhaps the docstring here is not [properly formatted](https://docs.ultralytics.com/help/contributing/#google-style-docstrings)? I noticed this line from the CI

```
WARNING -  griffe: ultralytics/data/augment.py:1452: Failed to get 'name: description' pair from 'Must be between 0 and 1, default is 0.0.'
```

Comment (ArgoHA) [reply ultralytics/data/augment.py]: that helped, thanks!

**Comment by ArgoHA** [comment]
@Burhan-Q so now we are just waiting for a review?

**Comment by Burhan-Q** [comment]
@ArgoHA yes I believe so. Make sure to keep the branch updated with changes from `main` and to ensure to resolve any conflicts or CI tests failures that _might_ occur. 🚀

**Comment by ArgoHA** [comment]
@Burhan-Q What is the typical waiting time? And are we sure I just need to wait?

**Comment by Burhan-Q** [comment]
> What is the typical waiting time? And are we sure I just need to wait?

@ArgoHA I don't think there's any "typical" waiting time. There's a lot going on and limited resources to allocate to everything. Contributions are _always_ welcome and appreciated, but there's never any assurance that a pull request will be reviewed or accepted. Even my pull requests can end up waiting for a while before getting a review, so don't take it personally, it's just a matter of the right person having the opportunity to review. 

Key things are to always keep the PR updated with main, make sure you've updated all relevant documentation pages, make sure all functions/classes include docstrings (with the correct format), write unit tests when applicable, provide a performance report when relevant, and check for any errors/typos. Aside from that, not much else I'm aware of that can help. 

For what it's worth, I'm trying to put together a process internally to help with the PR review process. It will take some time to get working if it's implemented, but I'm hoping it'll help.

**Comment by ArgoHA** [comment]
@Burhan-Q Hey, I believe there is again the same issue on your side during the `Ultralytics CI / Docs` test

**Comment by Burhan-Q** [comment]
Hm, strange. Not sure what the docs CI issue is (or how it got resolved). I noticed the Windows CI tests failed as well, so I restarted that. I'll have to ask about the docs CI

**Comment by ArgoHA** [comment]
Windows CI started failing 2 days ago for some reason. I obviously didn't change anything for a month or two already

**Comment by pderrenger** [comment]
Thank you for bringing this to our attention. The Windows CI failures might be due to recent updates or changes in dependencies. Could you please ensure that you are using the latest versions of the Ultralytics package and its dependencies? If the issue persists, providing any error logs or additional context would be helpful for further investigation.

**Comment by Burhan-Q** [comment]
Windows CI _might_ be a new Pytorch issue.

**Comment by ArgoHA** [comment]
@Burhan-Q Miscosoft problem disappeared, CI / Docs is still in place. I believe you already saw that problem last time:

> @ArgoHA sorry I missed that you had the permissions already set, that was a new problem I was seeing in a handful of PRs so I was taking a stab at what might help. I think the problem may have been resolved later yesterday.

Maybe you know how to resolve it again? It's exactly the same error: [old](https://github.com/ultralytics/ultralytics/actions/runs/9450175687/job/26028142922), [new](https://github.com/ultralytics/ultralytics/actions/runs/10124807598/job/27999582089?pr=13349)

**Comment by Burhan-Q** [comment]
@ArgoHA I'm not sure on this one. I don't do a lot of work on the CI stuff personally, so I'm not terribly familiar with it. I looked at the changes to the workflow file, but didn't see anything drastic recently that seemed like it could cause this. 

One thing that comes to mind is branch protection. Wondering if there's something about using the `main` branch on your fork?

**Comment by Burhan-Q** [comment]
REFENCE: On a quick search, I found this issue https://github.com/ad-m/github-push-action/issues/96 that might point to some things (for the CI configuration) that could help.

**Comment by ArgoHA** [comment]
@Burhan-Q I believe we have already [discussed that](https://github.com/ultralytics/ultralytics/pull/13349#issuecomment-2159820163). I have all the settings as it is in the link you attached. And I want to reiterate - I didn't change anything for a month+ and that test just stopped failing. Are you sure it is not the same [as before](https://github.com/ultralytics/ultralytics/pull/13349#issuecomment-2161361407)?

**Comment by Burhan-Q** [comment]
@ArgoHA I added that link as reference for anyone who's able to look into this with more in depth understanding of the GitHub Actions. Sorry for the confusion there, I know we looked at the settings before, but there were some other comments on that issue about the workflow setup I was trying to leave as a starting point for anyone who might get around to looking into this. What is strange to me is that there are other PRs that are not failing with the same error #14344 for example, which will make it harder to pinpoint the source of the issue.

**Comment by glenn-jocher** [comment]
@ArgoHA hi, thanks for your patience. The team is aware of the PR, and it's in the review queue. We appreciate your contribution and will update you as soon as possible.

**Comment by ArgoHA** [comment]
Hey @glenn-jocher . This MR has been up for nearly 6 months, is there any timeline for a review?
I just see on my third dataset that this augmentation gives a measurable boost. I think it would be great to have it merged so everyone can use it. Let me know if you need anything else from my side and if you have a pipeline with MRs to review, can I get any timelines?

**Comment by Y-T-G** [comment]
I think this would be closed in favor of https://github.com/ultralytics/ultralytics/pull/18246 which is more versatile and should also support this with `RandomRotate90`
